src/webgpu/api/operation/vertex_state.spec.ts

```diff
@@ -0,0 +1,366 @@
+      },
+    });
+  }
+
+  CreateIndexBuffer(indexArray: Uint16Array | Uint32Array, indexFormat: GPUIndexFormat): GPUBuffer {
+    const indexBuffer = this.device.createBuffer({
+      size: indexArray.byteLength,
+      usage: GPUBufferUsage.INDEX,
+      mappedAtCreation: true,
+    });
+    if (indexFormat === 'uint16') {
+      new Uint16Array(indexBuffer.getMappedRange()).set(indexArray);
+    } else {
+      new Uint32Array(indexBuffer.getMappedRange()).set(indexArray);
+    }
+    indexBuffer.unmap();
+    return indexBuffer;
+  }
+
+  run(
+    indexBuffer: GPUBuffer,
+    indexCount: number,
+    indexFormat: GPUIndexFormat,
+    indexOffset: number = 0,
+    primitiveTopology: GPUPrimitiveTopology = 'triangle-list'
+  ): GPUBuffer {
+    let pipeline: GPURenderPipeline;
+    // The indexFormat must be set in render pipeline descriptor that specifys a strip primitive
+    // topology for primitive restart testing
+    if (primitiveTopology === 'line-strip' || primitiveTopology === 'triangle-strip') {
+      pipeline = this.MakeRenderPipeline(primitiveTopology, indexFormat);
+    } else {
+      pipeline = this.MakeRenderPipeline(primitiveTopology);
+    }
+
+    const colorAttachment = this.device.createTexture({
+      format: kTextureFormat,
+      size: { width: kWidth, height: kHeight, depth: 1 },
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.OUTPUT_ATTACHMENT,
+    });
+
+    const result = this.device.createBuffer({
+      size: byteLength,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    const vertexArray = new Float32Array(
+      /* prettier-ignore */ [
+      // float4 position
+      -1.0, 1.0, 0.0, 1.0,
+      1.0, -1.0, 0.0, 1.0,
+      1.0, 1.0, 0.0, 1.0,
+      -1.0, -1.0, 0.0, 1.0]
+    );
+
+    let bufferSize = vertexArray.byteLength;
+    // Create a large enough buffer for uint16 primitive restart value used as a big index.
+    if (
+      indexFormat === 'uint16' &&
+      primitiveTopology !== 'line-strip' &&
+      primitiveTopology !== 'triangle-strip'
+    ) {
+      bufferSize = 0xffff * 16;
+    }
+
+    const vertexBuffer = this.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.VERTEX,
+      mappedAtCreation: true,
+    });
+    new Float32Array(vertexBuffer.getMappedRange()).set(vertexArray);
+    vertexBuffer.unmap();
+
+    const encoder = this.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: colorAttachment.createView(),
+          loadValue: { r: 0.0, g: 0.0, b: 0.0, a: 1.0 },
+          storeOp: 'store',
+        },
+      ],
+    });
+    pass.setPipeline(pipeline);
+    pass.setVertexBuffer(0, vertexBuffer);
+    pass.setIndexBuffer(indexBuffer, indexFormat, indexOffset);
+    pass.drawIndexed(indexCount, 1, 0, 0, 0);
+    pass.endPass();
+    encoder.copyTextureToBuffer(
+      { texture: colorAttachment, origin: { x: 0, y: 0, z: 0 } },
```

Comment:
nit: origin default to {1, 1, 1} so it can just be {}

Reply:
The origin defaults to `{}` so it can be omitted entirely.